### ISSUE TITLE
IMAP in Localhost - Problem with Virtualmin Driver and plugins

### ISSUE THREAD

**Issue by diegoweb** [roundcube/roundcubemail#6489]
Hey!

I have verified that if you are running Roundcube with localhost settings (imap), everything that needs the domain to work properly, fails (when running in Virtualmin environment)

For example:
The Password Plugin with Virtualmin doesn't work properly when running Roundcube in localhost.
This part of the virtualmin driver:
`exec("$curdir/chgvirtualminpasswd modify-user --domain $domain --user $username --pass $newpass", $output, $returnvalue);
`
Would work as follow:
$domain = localhost
$username = user-domain.

The domain itself is wrong, it should use the domain filled in the login box.
The username is also wrong since in the config file it configured as below:
```
// Virtualmin Driver options
// -------------------------
// Username format:
// 0: username@domain
// 1: username%domain
// 2: username.domain
// 3: domain.username
// 4: username-domain
// 5: domain-username
// 6: username_domain
// 7: domain_username
$config['password_virtualmin_format'] = 4;
```
In some aspects this config is wrong, but it only work when configured in this way. The virtualmin itself is configured with the pattern user@domain, but Roundcube refuses to use this format. If I set this driver file to echo the username and domain in this part of the virtualmin driver, I can "debug" how the things are getting filled in the exec command:

```
if ($returnvalue == 0) {
            return PASSWORD_SUCCESS;
        }
        else {
            rcube::raise_error(array(
                'code' => 600,
                'type' => 'php',
                'file' => __FILE__, 'line' => __LINE__,
                'message' => "Password plugin: Unable to execute $curdir/chgvirtualminpasswd"
                ), true, false); echo "Domain: ". $domain; echo " Username: ". $domain;
        }
```

If I go to the password change page and look at the header, it prints:
Domain: 'localhost' Username: 'user-domain'

So I had to change in all plugins the way they use the domain part and the username string.

This is my Virtualmin driver, maybe you could use some "if-clauses" in the future to detect if using localhost IMAP (running in Virtualmin) and set in the global config a new way to send the domain and the correct username for the whole roundcube script (so all old plugins would work properly):

```
<?php

/**
 * Virtualmin Password Driver
 *
 * Driver that adds functionality to change the users Virtualmin password.
 * The code is derrived from the Squirrelmail "Change Cyrus/SASL Password" Plugin
 * by Thomas Bruederli.
 *
 * It only works with virtualmin on the same host where Roundcube runs
 * and requires shell access and gcc in order to compile the binary.
 *
 * @version 3.0
 * @author Martijn de Munnik
 *
 * Copyright (C) 2005-2013, The Roundcube Dev Team
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see http://www.gnu.org/licenses/.
 */

class rcube_virtualmin_password
{
    function save($currpass, $newpass)
    {
        $rcmail   = rcmail::get_instance();
        $format   = $rcmail->config->get('password_virtualmin_format', 0);
        $username = $_SESSION['username'];
		$username = str_replace("-", "@", $username);

        switch ($format) {
        case 1: // username%domain
            $domain = substr(strrchr($username, "%"), 1);
            break;
        case 2: // username.domain (could be bogus)
            $pieces = explode(".", $username);
            $domain = $pieces[count($pieces)-2]. "." . end($pieces);
            break;
        case 3: // domain.username (could be bogus)
            $pieces = explode(".", $username);
            $domain = $pieces[0]. "." . $pieces[1];
            break;
        case 4: // username-domain
            $domain = substr(strrchr($username, "-"), 1);
            break;
        case 5: // domain-username
            $domain = str_replace(strrchr($username, "-"), "", $username);
            break;
        case 6: // username_domain
            $domain = substr(strrchr($username, "_"), 1);
            break;
        case 7: // domain_username
            $pieces = explode("_", $username);
            $domain = $pieces[0];
            break;
        default: // username@domain
            $domain = substr(strrchr($username, "@"), 1);
        }

        if (!$domain) {
            // $domain = $rcmail->user->get_username('domain'); // This causes error if running roundcube in localhost ($config['default_host'] = 'localhost' in roundcube/config/config.inc.php)
			$domain = str_replace("www.", "", $_SERVER['HTTP_HOST']);
        }
		
		$username = substr($username, 0, strrpos($username, '@')); // Remove @domain.tld from Username
        $username = escapeshellarg($username);
        $domain   = escapeshellarg($domain);
        $newpass  = escapeshellarg($newpass);
        $curdir   = RCUBE_PLUGINS_DIR . 'password/helpers';

        exec("$curdir/chgvirtualminpasswd modify-user --domain $domain --user $username --pass $newpass", $output, $returnvalue);

        if ($returnvalue == 0) {
            return PASSWORD_SUCCESS;
        }
        else {
            rcube::raise_error(array(
                'code' => 600,
                'type' => 'php',
                'file' => __FILE__, 'line' => __LINE__,
                'message' => "Password plugin: Unable to execute $curdir/chgvirtualminpasswd"
                ), true, false);z
        }

        return PASSWORD_ERROR;
    }
}

```

By the way, since I will install the Roundcube Webmail for each domain, I would rather lock the domain part so the user doesn't need to fill in the complete domain every time they want to log into the webmail. So my Global Roundcube Config looks like this:

```
// Automatically add this domain to user names for login
// Only for IMAP servers that require full e-mail addresses for login
// Specify an array with 'host' => 'domain' values to support multiple hosts
// Supported replacement variables:
// %h - user's IMAP hostname
// %n - hostname ($_SERVER['SERVER_NAME'])
// %t - hostname without the first part
// %d - domain (http hostname $_SERVER['HTTP_HOST'] without the first part)
// %z - IMAP domain (IMAP hostname without the first part)
// For example %n = mail.domain.tld, %t = domain.tld
$config['username_domain'] = '%d';

// Force domain configured in username_domain to be used for login.
// Any domain in username will be replaced by username_domain.
$config['username_domain_forced'] = true;
```

I had to manually change all plugins that require domain and username fields, for example the mfreiholz/persistent_login.

I don't know if my approach to adopt localhost imap I would be more secure since I don't need to "remote access" my imap server (the mail server is the same as my apache server, so everything is in the same server, localhost seems to be safer, just like when installing PHPMyAdmin and configuring it to run the mysql in localhost mode).

Could you in the next builds ask the user in the Global Config if running in Virtualmin environment and how the domain and username should be treated? This could easily fix all approaches, either localhost or remote imap, since we can use the %d variable and maybe a better way to str_replace the username-domain part. :)

**Comment by alecpl** [comment]
'default_host' also supports `%d` variable. I don't really see how we could improve here.

**Comment by diegoweb** [comment]
But if I set default_host to %d, my roundcube will not work in localhost anymore, right?
I mean, there will be packets sent to the internet, while in localhost everything is local. Or am I wrong and although set to %d, the server tries local queries first?

I just wanna improve the security as best as I can. Localhost mode seems to be a better choice for this purpose.

**Comment by alecpl** [comment]
Ok. Maybe the new 'password_username_format' option added in 60922dc3d5, set to `%d` would work for you? Or maybe you should just use 'mail_domain' setting?

Now as I look at it, I think setting 'username_domain' should work too, but problem is it works only for new users.

**Comment by diegoweb** [comment]
Well... I did a fresh install of my Virtualmin in a CentOS 7 machine and verified that Virtualmin is proper configured to use user@domain. I don't know why Roundcube use user-domain instead.

![image](https://user-images.githubusercontent.com/12632119/47304256-6fdc8d00-d5fc-11e8-8cce-6103f486e2f9.png)

![image](https://user-images.githubusercontent.com/12632119/47304400-ebd6d500-d5fc-11e8-83b9-9e4c9e9284fa.png)

OK! Nice. I have reverted the virtualmin driver in the password plugin to default.
For the mail_domain it worked, but the user keeps using the incorrect string, it should be removed by str_replace somewhere. This is how it is printed in the password page header:
Domain: 'domain.com' Username: 'user-domain.com'

I have tried both:
```
$config['password_virtualmin_format'] = 0;
$config['password_virtualmin_format'] = 4;
```

I have set this part of virtualmin driver to this:
```
else {
            rcube::raise_error(array(
                'code' => 600,
                'type' => 'php',
                'file' => __FILE__, 'line' => __LINE__,
                'message' => "Password plugin: Unable to execute $curdir/chgvirtualminpasswd"
                ), true, false); echo "Domain: ". $domain; echo " Username: ". $username; echo " Session Username: ". $_SESSION['username'];
        }
```
And as I can see, since there is no str_replace in the virtualmin driver, it prints user-domain.com.
There should be a `switch ($format) {` for username too, since there is the above 'password_virtualmin_format'. 
Domain: 'domain.com' Username: 'user-domain.com' Session Username: user-domain.com

But this would work only for the password plugin. Actually, the best approach should be a new "user string" in the global config, just like 'mail-domain'. I think this is the new 'password_username_format' you have mentioned, right? If so, I just need to download directly from roundcube website the version 1.4 Beta to test?
